### DISCUSSION TITLE
Need help sending 1 hex byte at a time through UART.

### DISCUSSION THREAD

**Discussion by donmerchant** [micropython/micropython#10925] [RP2040 / Pico]
I'm working on DMX with Raspberry Pico. I have a pico board with RS485 hat and trying to communicate with a Chauvet RGB T6 par light. I've tried some code from Jostein Lower [https://github.com/jostlowe/Pico-DMX](https://github.com/jostlowe/Pico-DMX) and Clacktronics [https://github.com/clacktronics/pyb_dmx](https://github.com/clacktronics/pyb_dmx) but can't get either of them to work. 

I have also programmed a Parallax Propeller 1 and gotten it to work and also have an ADJ DMX Controller that also works that I use as a reference. So I connected my Saleae logic analyzer to see what the output looks like. The only difference I can see between the Propeller / ADJ and the Pico programs is that on the Propeller and ADJ traces there is a MARK TIME AFTER SLOT after each byte sent in the packet that is not present in either of the Pico programs.

Josteins program is written in PIO and C (?) which I'm not educated enough to try and modify. So I've been playing with the Clacktronics version trying to send just 1 byte at a time and have the MARK TIME AFTER SLOT between each. No luck.

The problem I'm having is being able to send just 1 hex byte. For some reason I can't get it to be a hex number. Here's some code I tried:
```
from machine import UART, Pin
import time
from array import array

dmx_message = array('B', [0] * 9) #513)
dmx_message[1] = (0xFF)
dmx_message[2] = (0x80)
dmx_message[3] = (0xA0)
dmx_message[4] = (0x10)
dmx_message[5] = (0x55)

dmx_uart = Pin((4), Pin.OUT)
dmx_uart.value(0)
time.sleep_us(74)
dmx_uart.value(1)

dmx_uart = UART(1)

dmx_uart.init(250000, bits=8, parity=None, stop=2)
dmx_uart.write(hex(dmx_message[0]))
dmx_uart.deinit()

time.sleep_ms(10)

dmx_uart.init(250000, bits=8, parity=None, stop=2)
dmx_uart.write(hex(dmx_message[1]))
dmx_uart.deinit()

time.sleep_ms(10)

dmx_uart.init(250000, bits=8, parity=None, stop=2)
dmx_uart.write(hex(dmx_message[2]))
dmx_uart.deinit()
```
Maybe someone here can help me out understanding what I need to do?
Here's a screenshot showing the MARK between the stop of one byte and start of the next.
<img width="1399" alt="Screen Shot 2023-03-04 at 11 30 17 AM" src="https://user-images.githubusercontent.com/1148343/222920661-8d5c981c-2b71-4600-a11c-ea092fcc98ee.png">

**Comment by donmerchant** [comment]
Sorry those dark large commands are commented out in my code and showed up here for some reason.

**Reply by scruss** [reply]
please post you code with a line of three backticks (*```*) above and below and it will come out as code. What you've posted is impossible to follow

**Comment by karfas** [comment]
I have no experience using DMX at all, but shouldn't you send **binary** data over the UART ?
Your code `dmx_uart.write(hex(dmx_message[2]))` converts the byte to a hex string including "0x" (e.g. "0x25").

**Comment by donmerchant** [comment]
I found my answer after much searching. As mentioned above, yes it would need to be sent as binary data.

As an example this works:
```
uart.write(b'\x80')
uart.write(b'\xFF')
```
Question solved.

**Reply by Josverl** [reply]
UARTS have a difficult time adjusting to the DMX timings. It's not impossible but very MCU dependant.
On a RP2040 I think the best timing would be using the PIO.

Also, likely you'll need to increase the wait times between frames to significantly more than the paper spec. Cheap DMX equipment often is slower than the spec, and professional DMX.  Controllers slow down to deal with that

**Reply by Josverl** [reply]
This is proberen best writeup up I have seen on DMX

It is based on the ESP 32 and I2S
And it uses that in an interesting way to send well timed DMX signals.

https://github.com/someweisguy/esp_dmx
